src/platform/windows/misc.cpp

```diff
@@ -200,101 +202,84 @@ namespace platf {
+    DWORD consoleSessionId;
+    HANDLE userToken, duplicateToken;
+    TOKEN_ELEVATION_TYPE elevationType;
+    DWORD dwSize;
+
+    // Get the session ID of the active console session
+    consoleSessionId = WTSGetActiveConsoleSessionId();
+    if (0xFFFFFFFF == consoleSessionId) {
+      // If there is no active console session, log a warning and return null
+      BOOST_LOG(warning) << "There isn't an active user session, therefore it is not possible to execute commands under the users profile.";
+      return nullptr;
     }
 
-    // Duplicate the token to make it usable for process creation
-    HANDLE new_token;
-    ret = DuplicateTokenEx(shell_token, TOKEN_ALL_ACCESS, NULL, SecurityImpersonation, TokenPrimary, &new_token);
-    CloseHandle(shell_token);
-    if (!ret) {
-      BOOST_LOG(error) << "Failed to duplicate shell process token: "sv << GetLastError();
-      return NULL;
+    // Get the user token for the active console session
+    if (!WTSQueryUserToken(consoleSessionId, &userToken)) {
+      BOOST_LOG(debug) << "QueryUserToken failed, this would prevent commands from launching under the users profile.";
+      return nullptr;
     }
 
-    return new_token;
-  }
+    // Close the userToken handle when it goes out of scope
+    auto token_close = util::fail_guard([&userToken]() {
+      CloseHandle(userToken);
+    });
 
-  PTOKEN_USER
-  get_token_user(HANDLE token) {
-    DWORD return_length;
-    if (GetTokenInformation(token, TokenUser, NULL, 0, &return_length) || GetLastError() != ERROR_INSUFFICIENT_BUFFER) {
-      auto winerr = GetLastError();
-      BOOST_LOG(error) << "Failed to get token information size: "sv << winerr;
+    // We need to know if this is an elevated token or not.
+    // Get the elevation type of the user token
+    if (!GetTokenInformation(userToken, TokenElevationType, &elevationType, sizeof(TOKEN_ELEVATION_TYPE), &dwSize)) {
+      BOOST_LOG(debug) << "Retrieving token information failed: " << GetLastError();
       return nullptr;
     }
 
-    auto user = (PTOKEN_USER) HeapAlloc(GetProcessHeap(), 0, return_length);
-    if (!user) {
-      return nullptr;
-    }
+    if (elevated && elevationType == TokenElevationTypeDefault) {
+      // The token does not have a linked token, and user requested it to be elevated.
+      // This indicates that the user is not a member of the Administrators group.
+      BOOST_LOG(warning) << "This command requires elevation and the current user account logged in does not have administrator rights. "
+                         << "For security reasons Sunshine will retain the same access level as the current user and will not elevate it.";
 
-    if (!GetTokenInformation(token, TokenUser, user, return_length, &return_length)) {
-      auto winerr = GetLastError();
-      BOOST_LOG(error) << "Failed to get token information: "sv << winerr;
-      HeapFree(GetProcessHeap(), 0, user);
-      return nullptr;
+      return userToken;
```

Comment:
We can't return `userToken` here (or below in the failure case for GetTokenInformation) since our fail_guard action is going to close it. Can you refactor these 2 checks to fall-through on failure to call `DuplicateTokenEx()` will the current `userToken`?

Reply:
Actually, `WTSQueryUserToken()` returns a primary token, so I think we can drop the `DuplicateTokenEx()` call entirely and just directly return the handle we got from `WTSQueryUserToken()` or `GetTokenInformation()`. I would probably ditch the fail guard and just add `CloseHandle()` to the failure paths where it's required.